pkg/scheduler/plugins/reservation/plugin.go

```diff
@@ -0,0 +1,530 @@
+// Also do validations in this phase.
+func (p *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	// if the pod is a reserve pod
+	if IsReservePod(pod) {
+		// validate reserve pod and reservation
+		klog.V(3).InfoS("Attempting to pre-filter reserve pod", "pod", klog.KObj(pod))
+		r, err := p.lister.Reservations(pod.Namespace).Get(pod.Name)
+		if errors.IsNotFound(err) {
+			klog.Warningf("abort the pre-filter for reservation %s since it is not found", util.GetPodKey(pod))
+			return nil
+		} else if err != nil {
+			return framework.NewStatus(framework.Error, "cannot get reservation, err: "+err.Error())
+		}
+		err = validateReservation(r)
+		if err != nil {
+			return framework.NewStatus(framework.Error, err.Error())
+		}
+		return nil
+	}
+
+	klog.V(4).InfoS("Attempting to pre-filter pod for reservation state", "pod", klog.KObj(pod))
+	return nil
+}
+
+func (p *Plugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Filter only processes pods either the pod is a reserve pod or a pod can allocate reserved resources on the node.
+func (p *Plugin) Filter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	// the pod is a reserve pod
+	if IsReservePod(pod) {
+		klog.V(3).InfoS("Attempting to filter reserve pod", "pod", klog.KObj(pod), "node", node.Name)
+		// if the reservation specifies a nodeName initially, check if the nodeName matches
+		rNodeName := GetReservePodNodeName(pod)
+		if len(rNodeName) > 0 && rNodeName != nodeInfo.Node().Name {
+			return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonNodeNotMatchReservation)
+		}
+		// TODO: handle pre-allocation cases
+
+		return nil
+	}
+
+	// 2. check if the pod match a reservation
+	matchedCache := getPreFilterState(cycleState)
+	// expected matchedCache is prepared
+	if matchedCache == nil {
+		return framework.NewStatus(framework.Error, ErrReasonUnexpectedState)
+	}
+	return nil
+}
+
+func (p *Plugin) Score(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) (int64, *framework.Status) {
+	// if pod is a reserve pod, ignored
+	if IsReservePod(pod) {
+		return framework.MinNodeScore, nil
+	}
+
+	matchedCache := getPreFilterState(cycleState)
+	if matchedCache == nil { // expected matchedCache is prepared
+		return framework.MinNodeScore, framework.NewStatus(framework.Error, ErrReasonUnexpectedState)
+	}
+	rOnNode := matchedCache.GetOnNode(nodeName)
+	klog.V(4).InfoS("Attempting to score pod for reservation state", "pod", klog.KObj(pod),
+		"node", nodeName, "matched count", len(rOnNode))
+
+	// prefer the nodes which have reservation matched
+	if len(rOnNode) <= 0 {
```

Comment:
Here you can consider implementing a more refined scoring mechanism. If this version does not support it, it is recommended to add a TODO.

Reply:
done